### PR TITLE
Update the attention_mask reformat for MHA

### PR DESCRIPTION
**Pull request by apsonawane**
For phi3 models the attention_mask reformat was incorrect. Updated the pattern in this below PR.
This PR also helps with the issue open here: https://github.com/microsoft/onnxruntime-genai/issues/552

### CODE REVIEW

File: src/python/py/models/builder.py [hunk 1]
```diff
@@ -1689,61 +1694,73 @@ def make_attention_mask_reformatting_for_mha(self):
         # Make nodes for the attention mask subgraphs that reformat the
         # 2D attention mask (B, S) to 4D causal attention mask (B, N, S, T)
```
Comment (tianleiwu) [inline src/python/py/models/builder.py:1695]: I do not under stand this. I think the 2D mask is (B, T), and MHA supports 2D mask, shall we use that directly instead of converting to 4D in onnx graph?

**Review by natke — COMMENTED** [review]
Does this change get validated by tests?

**Comment by baijumeswani** [comment]
[apsonawane](https://github.com/apsonawane) is this pull-request still relevant? Could you please address the comments and update the PR when possible?

**Comment by apsonawane** [comment]
Yes, I will be updating the PR

**Comment by myadav2** [comment]
I've been experiencing a similar issue when fine-tuning phi3 and phi3.5 models as well. Produces a lot of tokens at the end that are gibberish even after fine-tuning. 

Looks like this has been solved with the latest ONNX release, but fine-tuning these ONNX models by converting them to torch is really tricky. Has the fix been made to the non-ONNX models as well? Any workaround for that?

**Comment by yufenglee** [comment]
> Looks like this has been solved with the latest ONNX release, but fine-tuning these ONNX models by converting them to torch is really tricky. Has the fix been made to the non-ONNX models as well? Any workaround for that?

@myadav2, the overall process for fine-tuning is that: fine-tune with Pytorch -> use [ModelBuilder](https://github.com/microsoft/onnxruntime-genai/tree/main/src/python/py/models) or PyTorch ONNX export to convert the model to ONNX -> serve with ORT GenAI API

**Comment by myadav2** [comment]
> > Looks like this has been solved with the latest ONNX release, but fine-tuning these ONNX models by converting them to torch is really tricky. Has the fix been made to the non-ONNX models as well? Any workaround for that?
> 
> @myadav2, the overall process for fine-tuning is that: fine-tune with Pytorch -> use [ModelBuilder](https://github.com/microsoft/onnxruntime-genai/tree/main/src/python/py/models) or PyTorch ONNX export to convert the model to ONNX -> serve with ORT GenAI API

Yes, but the issue of gibberish output after fine-tuning long context text is present for the base Pytorch models. I don't think we have a fix for that as far as I know

**Comment by yufenglee** [comment]
> > > Looks like this has been solved with the latest ONNX release, but fine-tuning these ONNX models by converting them to torch is really tricky. Has the fix been made to the non-ONNX models as well? Any workaround for that?
> > 
> > 
> > @myadav2, the overall process for fine-tuning is that: fine-tune with Pytorch -> use [ModelBuilder](https://github.com/microsoft/onnxruntime-genai/tree/main/src/python/py/models) or PyTorch ONNX export to convert the model to ONNX -> serve with ORT GenAI API
> 
> Yes, but the issue of gibberish output after fine-tuning long context text is present for the base Pytorch models. I don't think we have a fix for that as far as I know

@myadav2, that's annoying. Could you please describe your issue in details and share some examples? I can bring the issues to the Phi3 model training team and see if they can help.

**Comment by kunal-vaishnavi** [comment]
> Has the fix been made to the non-ONNX models as well? Any workaround for that?

The [Phi-3.5 PyTorch models](https://huggingface.co/collections/microsoft/phi-3-6626e15e9585a200d2d761e3) should also have this fix. If you still observe this behavior, you can open an issue in this repo so this can be tracked.
